runtime-modules/council/src/lib.rs

```diff
@@ -0,0 +1,719 @@
+// TODO: module documentation
+// TODO: adjust all extrinsic weights
+
+/////////////////// Configuration //////////////////////////////////////////////
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// used dependencies
+use codec::{Decode, Encode};
+use frame_support::traits::{Currency, Get, LockIdentifier, LockableCurrency, WithdrawReason};
+use frame_support::{decl_error, decl_event, decl_module, decl_storage, error::BadOrigin};
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+use std::marker::PhantomData;
+use system::{ensure_root, ensure_signed, RawOrigin};
+
+use referendum::Instance as ReferendumInstanceGeneric;
+use referendum::ReferendumManager;
+use referendum::Trait as ReferendumTrait;
+
+// declared modules
+mod mock;
+mod tests;
+
+/////////////////// Data Structures ////////////////////////////////////////////
+
+/// Information about council's current state and when it changed the last time.
+#[derive(Encode, Decode, PartialEq, Eq, Debug, Default)]
+pub struct CouncilStageInfo<AccountId, Currency, BlockNumber> {
+    stage: CouncilStage<AccountId, Currency>,
+    changed_at: BlockNumber,
```

Comment:
Relevant scenarios:
- when not enough candidates announced their candidacy -> restart announcement period
- when not enough council members were voted during voting/reveal period -> restart announcement period
- due to bad runtime implementation referendum's `start_referendum()` can fail during council's `end_announcement_period()`(it contains `.unwrap()` call). But only reason `start_referendum()` can fail is if referendum is already running - that should never happen with proper runtime, because council itself doesn't start referendum outside of expected period. There is currently no recovery from that and I'm not sure if can statically ensure it is implemented well.